Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/DaySyncTool.kt

```diff
@@ -0,0 +1,142 @@
+import de.rki.coronawarnapp.diagnosiskeys.storage.CachedKey
+import de.rki.coronawarnapp.diagnosiskeys.storage.CachedKeyInfo.Type
+import de.rki.coronawarnapp.diagnosiskeys.storage.KeyCacheRepository
+import de.rki.coronawarnapp.storage.DeviceStorage
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.toLocalDate
+import de.rki.coronawarnapp.util.TimeStamper
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.withContext
+import org.joda.time.LocalDate
+import timber.log.Timber
+import javax.inject.Inject
+
+@Reusable
+class DaySyncTool @Inject constructor(
+    deviceStorage: DeviceStorage,
+    private val keyServer: DiagnosisKeyServer,
+    private val keyCache: KeyCacheRepository,
+    private val downloadTool: KeyDownloadTool,
+    private val timeStamper: TimeStamper,
+    private val configProvider: AppConfigProvider,
+    private val dispatcherProvider: DispatcherProvider
+) : BaseSyncTool(
+    keyCache = keyCache,
+    deviceStorage = deviceStorage,
+    tag = TAG
+) {
+
+    /**
+     * returns true if the sync was successful
+     * and false if not all files have been synced
+     */
+    internal suspend fun syncMissingDays(
+        availableLocations: List<LocationCode>,
+        forceSync: Boolean
+    ): SyncResult {
+        Timber.tag(TAG).v("syncMissingDays(availableCountries=%s)", availableLocations)
+
+        val downloadConfig: KeyDownloadConfig = configProvider.getAppConfig()
+        invalidateCachedKeys(downloadConfig.invalidDayETags)
+
+        val missingDays = availableLocations.mapNotNull {
+            determineMissingDays(it, forceSync)
+        }
+        if (missingDays.isEmpty()) {
+            Timber.tag(TAG).i("There were no missing days.")
+            return SyncResult(successful = true, newPackages = emptyList())
+        }
+
+        Timber.tag(TAG).d("Downloading missing days: %s", missingDays)
+        requireStorageSpace(missingDays)
+
+        val downloads = launchDownloads(missingDays, downloadConfig)
+
+        Timber.tag(TAG).d("Waiting for %d missing day downloads.", downloads.size)
+        val downloadedDays = downloads.awaitAll().filterNotNull().also {
+            Timber.tag(TAG).v("Downloaded keyfile: %s", it.joinToString("\n"))
+        }
+        Timber.tag(TAG).i("Download success: ${downloadedDays.size}/${downloads.size}")
+
+        return SyncResult(
+            successful = downloads.size == downloadedDays.size,
+            newPackages = downloadedDays
+        )
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun expectNewDayPackages(cachedDays: List<CachedKey>): Boolean {
+        val yesterday = timeStamper.nowUTC.toLocalDate().minusDays(1)
+        val newestDay = cachedDays.map { it.info.createdAt }.maxOrNull()?.toLocalDate()
```

Comment:
Are we sure we can use createdAt to find the most recent Day Package? What if the user clears the cache, files are downloaded in bulk, and an older package created (stored) before a more recent package.

Reply:
Good point :+1:.

What do you think about this:
https://github.com/corona-warn-app/cwa-app-android/pull/1550/commits/700b3511da0db3e7c690cd28b994790c7060cefb#diff-e8b0c086fb3151818210311734b3595f9bf8fa8b304d188da76982fe3ea4ee52R53

Reply:
Yes, I think something like this is more reliable!